<a href="https://colab.research.google.com/github/angelajmzhou/ecs-171/blob/main/Lab1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lab1: Complete the TODO parts in the following code.
- Using California Housing Dataset from sklearn, select input attributes 1,3,4  as the input features.
- Using K-fold cross validation technique (https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html), complete the implementation to train a regression model and report performance merics when asked in the following code.
- For multiple degrees of model complexity (i.e., degree of polynomial in this exercise) in a for-loop, obtain the model with the minimum reducible_error, polynomial degree, and run the obtained model on the test data. For this part,you should use the split the data into train and test by [75:25] rate and report mse of the final model on test data.
- Analyse the results of model performance according to different degrees of polynomial and the number of folds used. You can manipulate the code and share your analysis in terms of the performance of the model (mse and total error), such as for instnace which degree of the model complexity (in relation to the polynomial order) would give a better model? Feel free to include other analysis about the generated models in relation to their performance results. You can event plot the results to support your analysis.

In [21]:
import numpy as np
from sklearn.model_selection import KFold
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import PolynomialFeatures
from sklearn.datasets import fetch_california_housing

def polynomial_regression(degree, X, y, folds, test_size=0.25, random_state=None):
    # Define number of folds for cross-validation
    kf = KFold(folds)

    # Initialize lists to store results for variance, bias2s, total_error, and models
    # TODO
    variance_list = []
    bias2s_list = []
    total_error = []
    models = []

    # Set the polynomial degree of the model
    poly_features = PolynomialFeatures(degree)
    X_poly = poly_features.fit_transform(X)

    # Perform cross-validation
    for train_index, test_index in kf.split(X_poly):
        # Split data into training and testing sets for this fold
        X_train, X_test = X_poly[train_index], X_poly[test_index]
        y_train, y_test = y[train_index], y[test_index]

        # Fit polynomial regression model
        model = LinearRegression()
        model.fit(X_train, y_train)

        # Make predictions on the test set
        y_pred = model.predict(X_test)

        # Calculate variance and Bias^2 for this fold
        variance = np.mean(((y_pred)-np.mean(y_pred))**2)
        bias2 = np.mean(((np.mean(y_pred)- y_test)) ** 2)

        # Append results to lists
        variance_list.append(variance)
        bias2s_list.append(bias2)
        total_error.append((bias2+variance))
        models.append(model)

        # Print results  including variance, Bias^2 and the total_error for this fold.
        print("Variance: {:.4f}, Bias2: {:.4f}, Total error: {:.4f}".format(variance, bias2, bias2 + variance))

    # print the total_error of the best model
    min_error_index = np.argmin(total_error)
    best_model = models[min_error_index]
    print("model number: {:1f} total error :{:4f}".format(min_error_index, total_error[min_error_index]))

    # Testing the final model on the test data
    X_train, X_test, y_train, y_test = train_test_split(X_poly, y, test_size=test_size, random_state=42)
    # Obtain the predictions on the test data
    best_model.fit(X_train, y_train)
    y_pred = best_model.predict(X_test)
    variance = np.mean(((y_pred)-np.mean(y_pred))**2)
    bias2 = np.mean(((np.mean(y_pred)- y_test)) ** 2)
    print("Best Model Values: Variance: {:.4f}, Bias2: {:.4f}, Total error: {:.4f}".format(variance, bias2, bias2 + variance))
    # Calculate the mean squared error (MSE) on the test data
    mse = mean_squared_error(y_test, y_pred)
    return mse , best_model






# Example usage: load California Housing Dataset and select the first, third, and forth attributes as input features in X
# TODO
housing_data = fetch_california_housing()
X = housing_data.data[:, [0, 2, 3]] #: means all rows, [0,2,3] means these specific columns
y = housing_data.target

# Set the target variable -- what goes into y, ie what we have to predict.

degrees = range(1, 6)  # Try polynomial degrees from 1 to 5

for degree in degrees:
    mse, best_model = polynomial_regression(degree, X, y, folds=5, random_state=42)
    print("Degree:", degree, "MSE:", mse)
# Try degrees from 1 to 5 and in a loop, report mse of the best model trained using k-fold cross validation and print("Degree:", degree, "MSE:", mse)


Variance: 0.6159, Bias2: 1.1592, Total error: 1.7751
Variance: 0.6857, Bias2: 1.2147, Total error: 1.9004
Variance: 0.8001, Bias2: 1.4449, Total error: 2.2450
Variance: 0.4341, Bias2: 1.1848, Total error: 1.6190
Variance: 0.7587, Bias2: 1.4722, Total error: 2.2309
model number: 3.000000 total error :1.618951
Best Model Values: Variance: 0.6780, Bias2: 1.3232, Total error: 2.0012
Degree: 1 MSE: 0.6668315328738224
Variance: 1.4212, Bias2: 1.1390, Total error: 2.5602
Variance: 0.6013, Bias2: 1.1999, Total error: 1.8012
Variance: 0.8027, Bias2: 1.4435, Total error: 2.2462
Variance: 0.5257, Bias2: 1.1828, Total error: 1.7085
Variance: 0.8059, Bias2: 1.4748, Total error: 2.2808
model number: 3.000000 total error :1.708497
Best Model Values: Variance: 1.0122, Bias2: 1.3234, Total error: 2.3356
Degree: 2 MSE: 0.915999395114274
Variance: 11.4957, Bias2: 1.1280, Total error: 12.6237
Variance: 0.5895, Bias2: 1.1955, Total error: 1.7849
Variance: 0.9781, Bias2: 1.4401, Total error: 2.4183
Variance

#ANALYSIS
**Analyse the results of model performance according to different degrees of polynomial and the number of folds used. You can manipulate the code and share your analysis in terms of the performance of the model (mse and total error),
such as for instnace which degree of the model complexity (in relation to the polynomial order) would give a better model? Feel free to include other analysis about the generated models in relation to their performance results.
You can event plot the results to support your analysis.**




One interesting pattern I noticed with the default k-fold validation function settings was that the first split in each degree would have significantly higher variance and bias compared to the subsequent splits, exponentially so for the higher degrees. However, when I set the shuffle parameter to true, that problem went away, indicating that the first segment of data had a significant amount of outliers. Although shuffling the folds helped with the first fold, it resulted in higher error throughout the other folds because the outliers were now distributed more evenly. Thus, I set shuffle to false again to prevent the outliers from affecting the other folds (although EDA would typically prune those values). Analyzing these results, the best model (purely by the metric of MSE) was the polynomial model of degree 3 [0.667], followed closely by the model of degree 1 [0.614]. However, looking at the metric of total error (variance + bias squared) calculated from the test split results, the degree 1 polynomial model slightly edged out the degree 3 polynomial model for the top spot, with a total error of 2.0012 compared to 2.0615. These results indicate that the relationship between the input data and target values is on the simpler side, and adding more degrees would lead to overfitting. Although the total error is lower for the one-degree model, the bias squared is the same for both, meaning the average model prediction is about the same distance from the ground truth. The difference between the two total errors thus solely originates from the variance, which only provides information on model complexity and the variability of its predictions, which naturally increases as the model becomes more complicated. Therefore, since the difference in total error can solely be attributed to the higher complexity of the three-degree model, with little bearing on the actual reliability of the model, the lower MSE of the three-degree model in comparison to the one-degree model leads me to believe that it is the best polynomial model for this dataset.

